### ISSUE TITLE
JSON-based workspaces don't detect or read target data layout?

### ISSUE THREAD

**Issue by whentojump** [rust-lang/rust-analyzer#14673]
## Question

Looks like JSON-based workspace doesn't have an equivalent of `target_layout` filed as in Cargo-based ones?

https://github.com/rust-lang/rust-analyzer/blob/237ffa3997250b8569ced477a5072310eb48c847/crates/project-model/src/workspace.rs#L70-L93

So when creating the crate graph here, we pass in an `Err`:

https://github.com/rust-lang/rust-analyzer/blob/237ffa3997250b8569ced477a5072310eb48c847/crates/project-model/src/workspace.rs#L622

May I know whether this is intentional, or perhaps at least a FIXME should be tagged at this line?

<br>

## Observed effects

#8813 won't work in JSON-based workspaces, e.g.

```rust
let _x = [true; 100];
  //^^ [bool; _] 
```

the length part is still `Unknown` thus displayed as an underscore.

<br>

## Possible solutions? (if we are to fix it)

* reuse this `target_data_layout::get()` function for JSON-based workspaces?

    https://github.com/rust-lang/rust-analyzer/blob/237ffa3997250b8569ced477a5072310eb48c847/crates/project-model/src/target_data_layout.rs#L15-L37

    By the way, in this function there are two attempts of issuing `rustc` command. It looks to me the only difference between them is to `cd <project root>` or not. Idk, will the existence of `Cargo.toml` in the current directory affect the result of `rustc +nightly -Z unstable-options --print target-spec-json`?

* derive the data layout information from this provided triple?

    https://github.com/rust-lang/rust-analyzer/blob/62c81d62932c458e23975191ecc124916be0b35e/crates/project-model/src/project_json.rs#L83

Thanks for your effort!

**Comment by HKalbasi** [comment]
We definitely want target data layout for rust-project.json based projects, and your possible solution sections looks like it is to the point. It would be great if you come with a PR.

> Idk, will the existence of Cargo.toml in the current directory affect the result of rustc +nightly -Z unstable-options --print target-spec-json?

I think you can provide the custom target with `--target` for it?

**Comment by whentojump** [comment]
Thanks for your reply!

>  come with a PR.

Sure. Let me try.

> > Idk, will the existence of Cargo.toml in the current directory affect the result of rustc +nightly -Z unstable-options --print target-spec-json?
>
> I think you can provide the custom target with --target for it?

Yeah I agree. It is also reflected in the snippet (the `cmd.args(["--target", target])` part).

But my point was, the snippet seemingly wants to extract certain information **from `Cargo.toml`**  with this line:

https://github.com/rust-lang/rust-analyzer/blob/237ffa3997250b8569ced477a5072310eb48c847/crates/project-model/src/target_data_layout.rs#L18

and if this first run of `rustc` fails, it launches a second run of `rustc` without `.current_dir()` configuration.

-- I am just confused about this design, and "why this function is currently only used for Cargo-based workspaces". Anyway, let me try it for JSON-based workspaces too.

**Comment by HKalbasi** [comment]
> But my point was, the snippet seemingly wants to extract certain information from Cargo.toml with this line:

I guess in case of cargo, the caller of this function won't provide the target (unless it is explicitly provided as a config) and it finds the target by running rustc in the directory of the crate. So for the `rust-project.json` you can just provide the config.

**Comment by Veykril** [comment]
> By the way, in this function there are two attempts of issuing rustc command. It looks to me the only difference between them is to cd <project root> or not. Idk, will the existence of Cargo.toml in the current directory affect the result of rustc +nightly -Z unstable-options --print target-spec-json?

That's a bug, the first one should be invoking `cargo` with `rustc` as its subcommand

**Comment by davidbarsky** [comment]
> Though for rust-project.json I'm not sure if we wanna attempt using cargo at all? Usually rust-project.json indicates a different workflow than standard cargo.

At least in my employer's case, it's possible to invoke `rustc` directly (technically `cargo` as well, since we store the full sysroot in the repo, but I think `rustc` is more correct).

**Comment by whentojump** [comment]
> the first one should be invoking cargo with rustc as its subcommand

I guess so according to the context :)

Then one follow-up question is: what _extra_ infomation can `cargo` provide regarding the target, compared to the plain `rustc`? My search only leads to [this](https://doc.rust-lang.org/cargo/reference/unstable.html#per-package-target) unstable config parameter.

**Comment by Veykril** [comment]
Running the command through cargo allows honoring target settings set through `.cargo/config.toml`. Running plain rustc will ignore that.

**Comment by whentojump** [comment]
Ahh I see. Thanks!

**Comment by davidbarsky** [comment]
> Yes, rustc is fine to do. The question is whether attempting to invoke cargo first makes sense, which I would argue it does not. AS usual, for non cargo based projects we should probably offer a way to set this in a different way/discover it differently.

Sorry about that! I misread what you were saying.
